### ISSUE TITLE
use_session_with_seed odd behavior

### ISSUE THREAD

**Issue by gsabot** [rstudio/keras#413]
I encountered an odd issue, possibly a bug, with
use_session_with_seed.  I've distilled my original problem to a much
smaller toy data set, attached, that illustrates the issue.  I'd
appreciate any advice, workarounds, etc.  

If I start R and call use_session_with_seed() first thing in my
session to set the seed, not disabling parallelism, then my first
model build and retrain works fine.  I can call
use_session_with_seed() again and this time disable parallelism, then
I can build/train again and get similar looking results.  That all works great.

However, if I start R and call use_session_with_seed() and disable
parallelism in that first call, all subsequent build/train attempts on
the model return numerically incorrect results, with the in-sample loss
and error both growing without bound to huge numbers!

The data set and the model function are in the attached zipped Rdata file.
The model function looks like this:

example.network <- function(input_shape) 
{
  model <- keras_model_sequential() %>%
    layer_flatten(input_shape = input_shape) %>%
    layer_dense(units = 100, activation = "relu") %>%
    layer_dense(units = 100, activation = "relu") %>%
    layer_dense(units = 100, activation = "relu") %>%
    layer_dense(units = 1)
  model %>% compile(
    optimizer = "rmsprop",
    loss = "mse",
    metrics = c("mae")
  )
}

And the data is small (obviously too small for that network, but I'm
just trying to boil down to a minimal test case):

> dim(train_data2)
[1] 200  21   1
> length(train_targets2)
[1] 200

Here is an example of what I believe to be "correct" behavior: I set
the seed, then build and train without shutting off parallelism, then
I do the same with parallelism turned off.  Both results happen to be
identical (I imagine turning off parallelism could have changed execution ordering
and modified things somewhat, but not in this case):
----------------
#Start fresh R and:
> load("/tmp/tr.RData")
> library("magrittr")
> library("keras")
> tensorflow:::use_session_with_seed(1,disable_gpu=FALSE,disable_parallel_cpu=FALSE)
/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
Set session seed to 1
> model <- do.call("example.network",list(c(21,1)))
> history <- model %>% fit( train_data2, train_targets2, epochs = 250, batch_size = 32, verbose = 0)
> head(as.data.frame(history$metrics));tail(as.data.frame(history$metrics))
          loss mean_absolute_error
1 1.011256e-04         0.005501779
2 1.801471e-06         0.001096648
3 2.968252e-06         0.001396685
4 1.697535e-06         0.001052120
5 6.385672e-06         0.002238516
6 1.945106e-06         0.001098141
            loss mean_absolute_error
245 1.423481e-06        0.0009412448
246 1.648461e-06        0.0010187984
247 1.325440e-06        0.0008878462
248 1.583438e-06        0.0009781588
249 1.690416e-06        0.0010566779
250 1.371779e-06        0.0009341308

#rerun, turn off parallelism (note, I have no GPU)
> tensorflow:::use_session_with_seed(1,disable_gpu=FALSE,disable_parallel_cpu=TRUE)
Set session seed to 1 (disabled CPU parallelism)
> model <- do.call("example.network",list(c(21,1)))
> history <- model %>% fit( train_data2, train_targets2, epochs = 250, batch_size = 32, verbose = 0)
> head(as.data.frame(history$metrics));tail(as.data.frame(history$metrics))
          loss mean_absolute_error
1 1.011256e-04         0.005501779
2 1.801471e-06         0.001096648
3 2.968252e-06         0.001396685
4 1.697535e-06         0.001052120
5 6.385672e-06         0.002238516
6 1.945106e-06         0.001098141
            loss mean_absolute_error
245 1.423481e-06        0.0009412448
246 1.648461e-06        0.0010187984
247 1.325440e-06        0.0008878462
248 1.583438e-06        0.0009781588
249 1.690416e-06        0.0010566779
250 1.371779e-06        0.0009341308
----------------

And now for the broken behavior:  The only code difference is this
one starts with disable_parallel_cpu=TRUE instead of FALSE.  The resulting loss seems
to grow without bound with each epoch:

----------------
#Start fresh R and:

> load("/tmp/tr.RData")
> library("magrittr")
> library("keras")
> tensorflow:::use_session_with_seed(1,disable_gpu=FALSE,disable_parallel_cpu=TRUE)
/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
Set session seed to 1 (disabled CPU parallelism)
> model <- do.call("example.network",list(c(21,1)))
> history <- model %>% fit( train_data2, train_targets2, epochs = 250, batch_size = 32, verbose = 0)
> head(as.data.frame(history$metrics));tail(as.data.frame(history$metrics))
          loss mean_absolute_error
1 1.127841e-04         0.006052550
2 1.865772e-06         0.001085329
3 5.445550e-06         0.001930966
4 4.279496e-06         0.001766649
5 1.045178e-05         0.002968265
6 4.807987e-06         0.001832147
        loss mean_absolute_error
245 4.464697            2.112942
246 4.641862            2.154476
247 4.828617            2.197343
248 5.030972            2.242956
249 5.246114            2.290396
250 5.431869            2.330591
----------------

After that, the R stays broken, giving those same results for each build/train, even if
you turn parallelism back on.  At least, I haven't found a workaround once an R has gotten into this state.

Here is my session info:

> sessionInfo()
R version 3.4.1 Patched (2017-09-01 r73173)
Platform: x86_64-pc-linux-gnu/x86_64 (64-bit)
Running under: Ubuntu 16.04.2 LTS

Matrix products: default
BLAS: /usr/lib/atlas-base/libf77blas.so.3.0
LAPACK: /usr/lib/atlas-base/atlas/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_US.iso-8859-1       LC_NUMERIC=C                   
 [3] LC_TIME=en_US.iso-8859-1        LC_COLLATE=en_US.iso-8859-1    
 [5] LC_MONETARY=en_US.iso-8859-1    LC_MESSAGES=en_US.iso-8859-1   
 [7] LC_PAPER=en_US.iso-8859-1       LC_NAME=C                      
 [9] LC_ADDRESS=C                    LC_TELEPHONE=C                 
[11] LC_MEASUREMENT=en_US.iso-8859-1 LC_IDENTIFICATION=C            

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] keras_2.1.5  magrittr_1.5

loaded via a namespace (and not attached):
 [1] compiler_3.4.1  R6_2.2.2        whisker_0.3-2   base64enc_0.1-3
 [5] Rcpp_0.12.12    reticulate_1.6  tensorflow_1.5  zeallot_0.1.0  
 [9] jsonlite_1.5    tfruns_1.3     

Thanks!

--gary
[tr.RData.zip](https://github.com/rstudio/keras/files/2040992/tr.RData.zip)

**Comment by gsabot** [comment]
I tried to convert my test case above into a workaround, where I would basically run my test case at the start of R to innoculate the R against the problem.  I learned that I had misinterpreted some of what my example was showing.

Digging deeper, I now see that my first call to use_session_with_seed() is able to turn parallelism on or off as desired, but but subsequent calls (regardless of the new disable_parallel_cpu arg) are unable to change things.  (I may be misunderstanding how general use_session_with_seed() is meant to be; I assumed that you could call it multiple times in a single R session, am I wrong about that?)

In my first example above, the 2nd call had no effect other than setting the seed, and the second run ran in parallel despite the CPU arg being false (I can see the job getting say 165% CPU in top).  And in my second example, the *R* is nailed down to being serial, and being serial somehow is always growing the error without bound.

So now I think I have two issues:

. Is there a way to use_session_with_seed() a second time in an R session and have it obey altered CPU/GPU args?

. Why does in-sample error seem to grow without bound in the serial run, but not in the parallel run? I suspect that the problem may be in the recording of the metric, not in the model, because in my actual (non-toy) problem, the network is generating useful out-of-sample predictions despite the wacky metrics stored in the history.

**Comment by jjallaire** [comment]
Yes, there is only ever one TF session with seed per-process so the second call doesn't do anything.

No idea on the sample error issue.

**Comment by gsabot** [comment]
Thanks.  Is there a way to completely reset/restart the one TF session in my R process so that use_session_with_seed could change the enabled/disabled state without needing to exit/restart R itself? Being able to do that and switch back  and forth would make it  a bit easier to experiment and dig into the sample error issue.  I tried calling tensorflow:::tf$reset_default_graph() before my second call to use_session_with_seed, but still that second call does not do anything so that kind of graph reset might be necessary but alone it is not sufficient.

**Comment by gsabot** [comment]
I think I now understand the sample error issue, it is probably just exploding gradients.  How bad it gets is affected by batch size, with bigger batch sizes having less of a problem, and for any given batch size, parallelism also seems to help, as does gradient clipping, simplifying the network, etc.

So I think I'm set on the sample issue, but I'm still curious to hear if there is a way to reset TF/change parallel/serial multiple times without restarting R.

**Comment by jjallaire** [comment]
No, the mechanisms used by `tf_use_session_with_seed()` affect global initialization states for TF and CUDA, and the design of these systems is to initialize with them once at process startup.

**Comment by gsabot** [comment]
It looks like setting/resetting the random seed multiple times in a
session works ok and allows for repeated replicable experiments in a
single R session, so calling use_session_with_seed_cpu_val() more than
once with either the same or different seed values is fine (at least
on my CPU-only system, maybe not on GPU).

But it could be nice to catch the problem when a user (or their
evolving init code) inadvertently tries to change the CPU or GPU
setting more than once in a session, rather than the current behavior
where it silently ignores such requests.  There probably is a more
elegant way to do this, but what do you think of adding this type of
arg sanity check to use_session_with_seed():

#define/init at top level, just before use_session_with_seed():
.use_session_with_seed_prev_gpu_val <- list()
.use_session_with_seed_prev_cpu_val <- list()

#Then inside the body of use_session_with_seed, near the top, add this:
arg_check <- function(saved_val,new_val,val_name=deparse(substitute(new_val))){
   if(length(saved_val)>0
      && saved_val!=new_val) {
      stop(paste("Bad",val_name,"arg: it was already set to",
                 saved_val,
                 "earlier; no change is allowed for rest of this R session"))
   }
}
arg_check(.use_session_with_seed_prev_gpu_val,disable_gpu)
arg_check(.use_session_with_seed_prev_cpu_val,disable_parallel_cpu)
#save vals for later checking
.use_session_with_seed_prev_gpu_val <<- disable_gpu
.use_session_with_seed_prev_cpu_val <<- disable_parallel_cpu
#Optional: you could conditionally check for GPU and flag random seed change if
#re-setting that was not allowed on GPU
